Framework/Kernel/src/Unit.cpp

```diff
@@ -670,19 +672,51 @@ double dSpacing::singleFromTOF(const double tof) const {
     throw std::runtime_error("Cannot convert to d spacing. Quadratic doesn't "
                              "have a positive root");
   }
-  // and then solve quadratic using Muller formula
-  double sqrtTerm;
-  if (difa == 0) {
-    // avoid costly sqrt even though formula reduces to this
-    sqrtTerm = difc;
+
+  // --------------------
+  //  solve for dpsacing
+  // --------------------
+  // NOTE: stupid explicity is the beauty
+  static constexpr double EPSILON = std::numeric_limits<double>::epsilon();
+  if (std::abs(difa) < EPSILON) {
+    // TOF = DIFA * d^2 + DIFC * d + T0
+    // 0.0 = DIFA * d^2 + DIFC * d + (T0 - TOF)
+    // degenerates to
+    // TOF = DIFC * d + T0
+    // gives us
+    // d = (TOF - T0) / DIFC
+    //
+    // NOTE: DIFC is always positive!!!
+    return (tof - tzero) / difc;
   } else {
-    sqrtTerm = sqrt(difc * difc - 4 * difa * (tzero - tof));
+    // TOF = DIFA * d^2 + DIFC * d + T0
+    // 0.0 = DIFA * d^2 + DIFC * d + (T0 - TOF)
+    // solve for this quadratic equation, we have
+    //       -DIFC + SQRT(DIFC^2 - 4*DIFA*(T0 - TOF))
+    // d = -------------------------------------------
+    //                  2*DIFA
+    // and
+    //       -DIFC - SQRT(DIFC^2 - 4*DIFA*(T0 - TOF))
+    // d = -------------------------------------------
+    //                  2*DIFA
+    // NOTE:
+    //   bottom line is that d has to be a positive number that is closer to 0
+    double sqrtTerm = sqrt(difc * difc - 4 * difa * (tzero - tof));
+    double d1 = 0.5 / difa * (-difc + sqrtTerm);
+    double d2 = 0.5 / difa * (-difc - sqrtTerm);
+    double dspacing;
+    //
+    if (d1 * d2 < 0.0) {
+      dspacing = std::max(d1, d2);
+    } else {
+      d1 = std::max(0.0, d1);
```

Comment:
@wdzhou 
The edge case for two negative roots are handled above with the following checking
```c++
  if ((difa > 0) && ((tzero - tof) > 0)) {
    throw std::runtime_error("Cannot convert to d spacing. Quadratic doesn't "
                             "have a positive root");
  }
```

Reply:
Suggestion:  From https://github.com/mantidproject/mantid/blob/PD269_binningErrorInAlignAndFocusPowder/Framework/Kernel/src/Unit.cpp#L661 to https://github.com/mantidproject/mantid/blob/PD269_binningErrorInAlignAndFocusPowder/Framework/Kernel/src/Unit.cpp#L671, how about if () {} else if () {} else if () ... ?

Reply:
And also, it is better to add some doc for `DIFA < 0` such that in this case, there must at least 1 positive root?

Reply:
- For the edge cases, the current ones are functionally equivalent to the suggested if-elif structure, but cleaner. Therefore I don't believe it is necessary to make the change.
- For the second question regarding add docs on the sign of `DIFA`, I would advise again doing so as it might lead the future developers down the path that was causing us the trouble in the first place.  More specially, the associated issue was due to the fact that someone was adapting the solution based on the documentation of deprecated [AlignDetectors](https://docs.mantidproject.org/nightly/algorithms/AlignDetectors-v1.html), which used sign of `DIFA` to determine the root.